{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# from .out file to a plot of mutation directions\n",
    "\n",
    "put full path to single_locus_trajectories.out file, full path to reference .dat file, desired coverage cutoff, and desired minimum variant frequency below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "variant_path = \"/Users/e.smith.5/Documents/PhD/RSV_project/RSV_haplotype_reconstruction/data/Single_locus_trajectories05.out\"\n",
    "\n",
    "ref_dat_path = \"/Users/e.smith.5/Documents/PhD/RSV_project/RSV_haplotype_reconstruction/data/ref/Reading_Frames.dat\"\n",
    "\n",
    "coverage_threshold = 0\n",
    "\n",
    "minimum_frequency = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><div style = \"float: left;\"><span>295×4 DataFrame</span></div><div style = \"float: right;\"><span style = \"font-style: italic;\">270 rows omitted</span></div><div style = \"clear: both;\"></div></div><div class = \"data-frame\" style = \"overflow-x: scroll;\"><table class = \"data-frame\" style = \"margin-bottom: 6px;\"><thead><tr class = \"header\"><th class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">Row</th><th style = \"text-align: left;\">pos</th><th style = \"text-align: left;\">original_base</th><th style = \"text-align: left;\">variant_base</th><th style = \"text-align: left;\">mutation</th></tr><tr class = \"subheader headerLastRow\"><th class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\"></th><th title = \"Int64\" style = \"text-align: left;\">Int64</th><th title = \"String\" style = \"text-align: left;\">String</th><th title = \"String\" style = \"text-align: left;\">String</th><th title = \"CategoricalValue{String, UInt32}\" style = \"text-align: left;\">Cat…</th></tr></thead><tbody><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">1</td><td style = \"text-align: right;\">11</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">2</td><td style = \"text-align: right;\">18</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">3</td><td style = \"text-align: right;\">141</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">4</td><td style = \"text-align: right;\">144</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">G-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">5</td><td style = \"text-align: right;\">202</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">6</td><td style = \"text-align: right;\">380</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T-&gt;C</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">7</td><td style = \"text-align: right;\">561</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">8</td><td style = \"text-align: right;\">574</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T-&gt;C</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">9</td><td style = \"text-align: right;\">576</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">10</td><td style = \"text-align: right;\">754</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">11</td><td style = \"text-align: right;\">811</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">12</td><td style = \"text-align: right;\">977</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">C-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">13</td><td style = \"text-align: right;\">1007</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C-&gt;T</td></tr><tr><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">284</td><td style = \"text-align: right;\">14037</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">285</td><td style = \"text-align: right;\">14042</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">286</td><td style = \"text-align: right;\">14154</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">G-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">287</td><td style = \"text-align: right;\">14200</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">288</td><td style = \"text-align: right;\">14350</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">T-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">289</td><td style = \"text-align: right;\">14428</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T-&gt;C</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">290</td><td style = \"text-align: right;\">14431</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">A-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">291</td><td style = \"text-align: right;\">14552</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">292</td><td style = \"text-align: right;\">14616</td><td style = \"text-align: left;\">C</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">C-&gt;T</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">293</td><td style = \"text-align: right;\">14853</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A-&gt;G</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">294</td><td style = \"text-align: right;\">14968</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">A</td><td style = \"text-align: left;\">G-&gt;A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">295</td><td style = \"text-align: right;\">15041</td><td style = \"text-align: left;\">G</td><td style = \"text-align: left;\">T</td><td style = \"text-align: left;\">G-&gt;T</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& pos & original\\_base & variant\\_base & mutation\\\\\n",
       "\t\\hline\n",
       "\t& Int64 & String & String & Cat…\\\\\n",
       "\t\\hline\n",
       "\t1 & 11 & A & G & A->G \\\\\n",
       "\t2 & 18 & A & G & A->G \\\\\n",
       "\t3 & 141 & G & A & G->A \\\\\n",
       "\t4 & 144 & G & T & G->T \\\\\n",
       "\t5 & 202 & C & T & C->T \\\\\n",
       "\t6 & 380 & T & C & T->C \\\\\n",
       "\t7 & 561 & C & T & C->T \\\\\n",
       "\t8 & 574 & T & C & T->C \\\\\n",
       "\t9 & 576 & G & A & G->A \\\\\n",
       "\t10 & 754 & A & G & A->G \\\\\n",
       "\t11 & 811 & C & T & C->T \\\\\n",
       "\t12 & 977 & C & A & C->A \\\\\n",
       "\t13 & 1007 & C & T & C->T \\\\\n",
       "\t14 & 1009 & A & G & A->G \\\\\n",
       "\t15 & 1054 & C & T & C->T \\\\\n",
       "\t16 & 1056 & T & A & T->A \\\\\n",
       "\t17 & 1082 & G & A & G->A \\\\\n",
       "\t18 & 1084 & A & G & A->G \\\\\n",
       "\t19 & 1136 & C & T & C->T \\\\\n",
       "\t20 & 1257 & T & C & T->C \\\\\n",
       "\t21 & 1365 & C & A & C->A \\\\\n",
       "\t22 & 1471 & G & A & G->A \\\\\n",
       "\t23 & 1472 & C & T & C->T \\\\\n",
       "\t24 & 1579 & C & A & C->A \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m295×4 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0m│\u001b[1m pos   \u001b[0m\u001b[1m original_base \u001b[0m\u001b[1m variant_base \u001b[0m\u001b[1m mutation \u001b[0m\n",
       "     │\u001b[90m Int64 \u001b[0m\u001b[90m String        \u001b[0m\u001b[90m String       \u001b[0m\u001b[90m Cat…     \u001b[0m\n",
       "─────┼──────────────────────────────────────────────\n",
       "   1 │    11  A              G             A->G\n",
       "   2 │    18  A              G             A->G\n",
       "   3 │   141  G              A             G->A\n",
       "   4 │   144  G              T             G->T\n",
       "   5 │   202  C              T             C->T\n",
       "   6 │   380  T              C             T->C\n",
       "   7 │   561  C              T             C->T\n",
       "   8 │   574  T              C             T->C\n",
       "  ⋮  │   ⋮          ⋮             ⋮           ⋮\n",
       " 289 │ 14428  T              C             T->C\n",
       " 290 │ 14431  A              T             A->T\n",
       " 291 │ 14552  G              A             G->A\n",
       " 292 │ 14616  C              T             C->T\n",
       " 293 │ 14853  A              G             A->G\n",
       " 294 │ 14968  G              A             G->A\n",
       " 295 │ 15041  G              T             G->T\n",
       "\u001b[36m                                    280 rows omitted\u001b[0m"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read variants file into a dataframe\n",
    "using DataFrames\n",
    "\n",
    "# Define the function to read the .out file\n",
    "function read_out_file(filepath::String)\n",
    "    # Initialize an empty DataFrame\n",
    "    df = DataFrame(pos = Int[], original_base = String[], variant_base = String[])\n",
    "    \n",
    "    # Open the file and process each line\n",
    "    open(filepath, \"r\") do file\n",
    "        for line in eachline(file)\n",
    "            # Remove any leading/trailing whitespace\n",
    "            stripped_line = strip(line)\n",
    "            \n",
    "            # Continue only if the line is not empty\n",
    "            if !isempty(stripped_line)\n",
    "                # Split the line into parts, typically space-delimited\n",
    "                parts = split(stripped_line)\n",
    "                \n",
    "                # Ensure there are enough parts to avoid index errors\n",
    "                if length(parts) < 3\n",
    "                    continue  # Skip this line if it doesn't have enough parts\n",
    "                end\n",
    "\n",
    "                # Extract the position, original base, and variant base\n",
    "                pos = parse(Int, parts[1])\n",
    "                original_base = parts[2]  # Should be a single character string\n",
    "                variant_base = parts[3]   # Should be a single character string\n",
    "                \n",
    "                # Append to the DataFrame\n",
    "                push!(df, (pos, original_base, variant_base))\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return df\n",
    "end\n",
    "\n",
    "variants = read_out_file(variant_path)\n",
    "\n",
    "using CategoricalArrays\n",
    "\n",
    "function call_muts(df::DataFrame)\n",
    "    # Check if the required columns are present in the DataFrame\n",
    "    if !(\"original_base\" in names(df) && \"variant_base\" in names(df))\n",
    "        error(\"DataFrame must contain 'original_base' and 'variant_base' columns\")\n",
    "    end\n",
    "\n",
    "    # Create a new column 'mutation' by concatenating 'original_base', '->', 'variant_base'\n",
    "    mutations = df.original_base .* \"->\" .* df.variant_base\n",
    "\n",
    "    # Define the order of categories\n",
    "    levels_order = [\"A->C\", \"A->G\", \"A->T\", \"C->A\", \"C->G\", \"C->T\", \"G->A\", \"G->C\", \"G->T\", \"T->A\", \"T->C\", \"T->G\"]\n",
    "\n",
    "    # Convert 'mutations' into a categorical array with predefined levels\n",
    "    df.mutation = categorical(mutations, ordered=true, levels=levels_order)\n",
    "    \n",
    "    return df\n",
    "end\n",
    "\n",
    "variants = call_muts(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "RCall.REvalError",
     "evalue": "REvalError: Error in `ggplot()`:\n! `data` cannot be a function.\ni Have you misspelled the `data` argument in `ggplot()`\nBacktrace:\n    x\n 1. +-ggplot2::ggplot(data, aes(x = name, y = value))\n 2. \\-ggplot2:::ggplot.function(data, aes(x = name, y = value))\n 3.   \\-cli::cli_abort(c(\"{.arg data} cannot be a function.\", i = \"Have you misspelled the {.arg data} argument in {.fn ggplot}\"))\n 4.     \\-rlang::abort(...)",
     "output_type": "error",
     "traceback": [
      "REvalError: Error in `ggplot()`:\n",
      "! `data` cannot be a function.\n",
      "i Have you misspelled the `data` argument in `ggplot()`\n",
      "Backtrace:\n",
      "    x\n",
      " 1. +-ggplot2::ggplot(data, aes(x = name, y = value))\n",
      " 2. \\-ggplot2:::ggplot.function(data, aes(x = name, y = value))\n",
      " 3.   \\-cli::cli_abort(c(\"{.arg data} cannot be a function.\", i = \"Have you misspelled the {.arg data} argument in {.fn ggplot}\"))\n",
      " 4.     \\-rlang::abort(...)\n",
      "\n",
      "Stacktrace:\n",
      " [1] handle_eval_stderr(; as_warning::Bool)\n",
      "   @ RCall ~/.julia/packages/RCall/dDAVd/src/io.jl:174\n",
      " [2] handle_eval_stderr\n",
      "   @ ~/.julia/packages/RCall/dDAVd/src/io.jl:168 [inlined]\n",
      " [3] reval_p(expr::Ptr{LangSxp}, env::Ptr{EnvSxp})\n",
      "   @ RCall ~/.julia/packages/RCall/dDAVd/src/eval.jl:103\n",
      " [4] reval_p(expr::Ptr{RCall.ExprSxp}, env::Ptr{EnvSxp})\n",
      "   @ RCall ~/.julia/packages/RCall/dDAVd/src/eval.jl:119\n",
      " [5] reval(str::String, env::RObject{EnvSxp})\n",
      "   @ RCall ~/.julia/packages/RCall/dDAVd/src/eval.jl:136\n",
      " [6] macro expansion\n",
      "   @ ~/.julia/packages/RCall/dDAVd/src/macros.jl:71 [inlined]\n",
      " [7] top-level scope\n",
      "   @ ~/Documents/PhD/RSV_project/RSV_haplotype_reconstruction/src/variants.ipynb:5"
     ]
    }
   ],
   "source": [
    "# barplot of mutation frequencies\n",
    "using RCall\n",
    "\n",
    "@rput variants\n",
    "R\"\"\"\n",
    "library(ggplot2)\n",
    "# Barplot\n",
    "ggplot(data, aes(x=name, y=value)) +\n",
    "  geom_bar(stat = \"identity\")\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucleotide .fa : /Users/e.smith.5/Documents/PhD/RSV_project/RSV_haplotype_reconstruction/data/ref/Reading_Frames_nucleotide.fa\n",
      "amino acid .fa : /Users/e.smith.5/Documents/PhD/RSV_project/RSV_haplotype_reconstruction/data/ref/Reading_Frames_amino_acid.fa\n"
     ]
    }
   ],
   "source": [
    "# pull two .fas from .dat file \n",
    "using RCall\n",
    "\n",
    "# Function to handle the .dat file processing\n",
    "function pull_frames(dat_file_path::String)\n",
    "    # Define and call the R processing function within RCall\n",
    "    @rput dat_file_path  # This passes the Julia variable to R\n",
    "    R\"\"\"\n",
    "    library(Biostrings)\n",
    "\n",
    "    process_file <- function(file_path) {\n",
    "        lines <- tryCatch({\n",
    "            readLines(file_path)\n",
    "        }, error = function(e) {\n",
    "            cat(\"Error reading file:\", e$message, \"\\n\")\n",
    "            return(NULL)  # Return NULL to handle error gracefully\n",
    "        })\n",
    "        if (is.null(lines)) return()  # Exit if file could not be read\n",
    "\n",
    "        fasta_nucleotide_path <- sub(\".dat$\", \"_nucleotide.fa\", file_path)\n",
    "        fasta_amino_acid_path <- sub(\".dat$\", \"_amino_acid.fa\", file_path)\n",
    "        \n",
    "        con_nuc <- file(fasta_nucleotide_path, open = \"w\")\n",
    "        con_aa <- file(fasta_amino_acid_path, open = \"w\")\n",
    "        \n",
    "        for (i in seq(1, length(lines), by = 3)) {\n",
    "            if (i + 2 > length(lines)) {\n",
    "                break\n",
    "            }\n",
    "            header <- strsplit(lines[i], \" \")[[1]]\n",
    "            if (length(header) < 3) {\n",
    "                next\n",
    "            }\n",
    "            start_pos <- header[1]\n",
    "            end_pos <- header[2]\n",
    "            annotation <- paste(header[3:length(header)], collapse=\" \")\n",
    "            full_annotation <- paste0('\"accession\":', start_pos, '..', end_pos, ' | ', annotation)\n",
    "            dna_seq <- DNAString(lines[i+1])\n",
    "            aa_seq <- AAString(lines[i+2])\n",
    "            \n",
    "            writeLines(paste0(\">\", full_annotation), con_nuc)\n",
    "            writeLines(as.character(dna_seq), con_nuc)\n",
    "            writeLines(paste0(\">\", full_annotation), con_aa)\n",
    "            writeLines(as.character(aa_seq), con_aa)\n",
    "        }\n",
    "        \n",
    "        close(con_nuc)\n",
    "        close(con_aa)\n",
    "    }\n",
    "    \"\"\"\n",
    "    # Call the R function from Julia\n",
    "    R\"process_file($(dat_file_path))\"\n",
    "    \n",
    "    # Julia takes over after the R function to print the paths\n",
    "    fasta_nucleotide_path = replace(dat_file_path, \".dat\" => \"_nucleotide.fa\")\n",
    "    fasta_amino_acid_path = replace(dat_file_path, \".dat\" => \"_amino_acid.fa\")\n",
    "    \n",
    "    println(\"nucleotide .fa : \", fasta_nucleotide_path)\n",
    "    println(\"amino acid .fa : \", fasta_amino_acid_path)\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "pull_frames(ref_dat_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.3",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
